src/storage/pooled_storage_manager.h

```diff
@@ -89,16 +95,17 @@ void* GPUPooledStorageManager::Alloc(size_t size) {
   }
 }
 
-void GPUPooledStorageManager::Free(void* ptr, size_t size) {
+void GPUPooledStorageManager::Free(void* ptr, size_t raw_size) {
   std::lock_guard<std::mutex> lock(mutex_);
+  size_t size = raw_size + ndev_;
   auto&& reuse_pool = memory_pool_[size];
   reuse_pool.push_back(ptr);
 }
 
 void GPUPooledStorageManager::ReleaseAll() {
   for (auto&& i : memory_pool_) {
     for (auto&& j : i.second) {
-      DirectFree(j, i.first);
+      DirectFree(j, i.first - ndev_);
```

Comment:
This looks hacky. Are you using ndev_ bytes for storing some meta data? Also you didn't shift by +ndev_ in alloc